### PR TITLE
Bluetooth: controller: Fix missing storage for sync cancel reason

### PR DESCRIPTION
**Pull request by cvinayak**
Fix missing storage allocation in ll_sync_set structure for
storing the sync cancel reason.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:214]: in `ull_conn.c:1716` the node_rx is assigned to the outer struct, rather than the header. 

I think we should consider being consistent with that, so either do the same here, or change the `ull_conn.c` as well.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:218]: I'm not sure I understand this. 

`se` is of type `struct node_rx_sync` and `sync->node_rx_lost.reason` is just a `uint8_t`. By doing this, you make `status == reason`, but if trying to access any other member of the `se` struct, you access illegal data.

**Review by Thalley — CHANGES_REQUESTED** [review]

**Review by Thalley — APPROVED** [review]

**Comment by cvinayak** [comment]
@wopu-ot request your review. (do you think we should add you into code owners file, so that you are added automatically?)

File: subsys/bluetooth/controller/ll_sw/ull_master.c
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_master.c:411]: That's quite ugly and should be cleaned up. However, that should happen in a different PR (documenting ugliness should not be "rewarded" with scope creep).

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:220]: This is quite ugly. I would prefer to have the type of the `node_rx_lost` field be non-anonymous so `se` can be declared with that type and the `reason` field is accessible. I would also be okay with accessing `sync->node_rx_lost.reason` directly.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:176]: The compiler is likely smart enough that removing this auto variable does not save anything but only makes the code a little harder to read.

**Review by wopu-ot — CHANGES_REQUESTED** [review]

**Review by wopu-ot — COMMENTED** [review]
After the changes and re-reviewing, I think I understand a little more. Do I understand correctly that the `reason` field in `node_rx_lost` must be at the same offset as the `status` field in a `node_rx_pdu` with a `node_rx_sync` payload? A comment regarding this at the places where `node_rx_lost` and `node_rx_sync` are declared would be helpful to avoid future changes breaking this assumption.

Additionally, I realized that the `pdu` field in `node_rx_pdu` may be unaligned but `node_rx_sync` may have alignment requirements. Assignments like `struct node_rx_sync *se = (void *)node_rx->pdu` are therefore bound to break at some point. This does not necessarily need to be resolved in this PR, but should be resolved eventually.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:229]: I guess this is equivalent to writing
```
sync->node_rx_lost.hdr.type = NODE_RX_TYPE_SYNC;
sync->node_rx_lost.hdr.handle = 0xffff;
sync->node_rx_lost.hdr.rx_ftr.param = sync;
sync->node_rx_lost.reason = BT_HCI_ERR_OP_CANCELLED_BY_HOST;

*rx = &sync->node_rx_lost;
```
Isn't it? 

I wonder if this is more readable, as you're directly changing the values of the fields in the `sync->node_rx_lost`, rather than doing it through 2 other pointers. 

Similar comment also applies for the `ll_connect_disable` change above.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_sync.c:229]: The code is kept similar to HCI access to the reason, this way future code/features can follow the design.

**Comment by cvinayak** [comment]
>A comment regarding this at the places where `node_rx_lost` and `node_rx_sync` are declared would be helpful to avoid future changes breaking this assumption.

Will do.

> Additionally, I realized that the `pdu` field in `node_rx_pdu` may be unaligned but `node_rx_sync` may have alignment requirements. Assignments like `struct node_rx_sync *se = (void *)node_rx->pdu` are therefore bound to break at some point. 

Will address this in a separate PR.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/ull_sync.c:229]: I understand that, and will not block the PR based on that. 

However, if using the fields of the `sync->node_rx_lost` it would be possible, if ever needed, to expand that easier than if just expected a certain sequence of fields in that struct. But it's alright as it is right now.

File: subsys/bluetooth/controller/ll_sw/lll.h
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/lll.h:292]: I'm not sure I understand why the `__aligned(4)` part is necessary. Wouldn't the the `pdu[0]` fields be aligned correctly without it?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/lll.h:292]: `pdu[0]` is uint8_t and hence depending on the previous structure member size, would be placed unaligned (packed). HCI code typecasts unpacked structures to `&pdu[0]`, potentially leading to unaligned access faults if `struct node_rx_hdr` size changes.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/lll.h:292]: Alternatively, the below too solves the problem.

```suggestion
	union {
		uint8_t    pdu[0];
		uint32_t   param[0];
	};
```

Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/lll.h:292]: I see. Wouldn't defining pdu as a `void *` provide the same? Do we lose anything from doing that? 

It's not an issue that it's there, but I guess we'll be casting `pdu` to another struct definition anyhow, so does it matter that it's uint8_t or void?

Comment (wopu-ot) [reply subsys/bluetooth/controller/ll_sw/lll.h:292]: Defining pdu as void* would always occupy 4 bytes, even when not needed.

I think having the zero-length array with `__aligned(4)` is the correct solution. I am not sure whether the union in the `node_rx_pdu` struct is strictly necessary, but it is consistent with how the other struct look like, which is a good thing.

**Review by wopu-ot — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/lll.h
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/lll.h:292]: You are correct :) It could be `uint32_t pdu[0]` though, but I think that `__aligned(4)` serves the same purpose just as well.

**Pull request merged**
